File: app/src/main/java/com/xpe/mobile/editor/MathExpression.java
package com.xpe.mobile.editor;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/** Small dependency-free expression compiler used by the manual's Complex Move tool. */
public final class MathExpression {
    public interface Compiled {
        double evaluate(double t);
    }

    private interface Node {
        double evaluate(double t);
    }

    private static final int MAX_EXPRESSION_LENGTH = 1024;
    private static final int MAX_PARSE_DEPTH = 96;

    private MathExpression() {
    }

    public static Compiled compile(String source) {
        if (source == null || source.trim().isEmpty()) {
            throw new IllegalArgumentException("expression is required");
        }
        String normalized = source.replace("$", "").trim();
        if (normalized.length() > MAX_EXPRESSION_LENGTH) {
            throw new IllegalArgumentException("expression is too long");
        }
        Parser parser = new Parser(normalized);
        Node root = parser.parse();
        return t -> {
            if (!Double.isFinite(t)) throw new IllegalArgumentException("t must be finite");
            double value = root.evaluate(t);
            if (!Double.isFinite(value)) {
                throw new IllegalArgumentException("expression result is not finite");
            }
            return value;
        };
    }

    private static final class Parser {
        private final String source;
        private int position;
        private int depth;

        Parser(String source) {
            this.source = source;
        }

        Node parse() {
            Node node = parseAdditive();
            skipWhitespace();
            if (!atEnd()) throw error("unexpected character");
            return node;
        }

        private Node parseAdditive() {
            enter();
            try {
                Node node = parseMultiplicative();
                while (true) {
                    if (take('+')) {
                        Node left = node;
                        Node right = parseMultiplicative();
                        node = t -> left.evaluate(t) + right.evaluate(t);
                    } else if (take('-')) {
                        Node left = node;
                        Node right = parseMultiplicative();
                        node = t -> left.evaluate(t) - right.evaluate(t);
                    } else {
                        return node;
                    }
                }
            } finally {
                leave();
            }
        }

        private Node parseMultiplicative() {
            Node node = parseUnary();
            while (true) {
                if (take('*')) {
                    Node left = node;
                    Node right = parseUnary();
                    node = t -> left.evaluate(t) * right.evaluate(t);
                } else if (take('/')) {
                    Node left = node;
                    Node right = parseUnary();
                    node = t -> left.evaluate(t) / right.evaluate(t);
                } else if (take('%')) {
                    Node left = node;
                    Node right = parseUnary();
                    node = t -> left.evaluate(t) % right.evaluate(t);
                } else {
                    return node;
                }
            }
        }

        private Node parseUnary() {
            if (take('+')) return parseUnary();
            if (take('-')) {
                Node value = parseUnary();
                return t -> -value.evaluate(t);
            }
            return parsePower();
        }

        private Node parsePower() {
            Node base = parsePrimary();
            if (!take('^')) return base;
            Node exponent = parseUnary();
            return t -> Math.pow(base.evaluate(t), exponent.evaluate(t));
        }

        private Node parsePrimary() {
            skipWhitespace();
            if (take('(')) {
                Node value = parseAdditive();
                require(')');
                return value;
            }
            if (atEnd()) throw error("value expected");
            char current = source.charAt(position);
            if (Character.isDigit(current) || current == '.') return parseNumber();
            if (Character.isLetter(current) || current == '_') return parseIdentifier();
            throw error("value expected");
        }

        private Node parseNumber() {
            int start = position;
            boolean digit = false;
            while (!atEnd() && Character.isDigit(source.charAt(position))) {
                position++;
                digit = true;
            }
            if (!atEnd() && source.charAt(position) == '.') {
                position++;
                while (!atEnd() && Character.isDigit(source.charAt(position))) {
                    position++;
                    digit = true;
                }
            }
            if (!digit) throw error("invalid number");
            if (!atEnd() && (source.charAt(position) == 'e' || source.charAt(position) == 'E')) {
                int exponent = position++;
                if (!atEnd() && (source.charAt(position) == '+' || source.charAt(position) == '-')) {
                    position++;
                }
                int exponentStart = position;
                while (!atEnd() && Character.isDigit(source.charAt(position))) position++;
                if (position == exponentStart) {
                    position = exponent;
                    throw error("invalid exponent");
                }
            }
            double value;
            try {
                value = Double.parseDouble(source.substring(start, position));
            } catch (NumberFormatException exception) {
                throw error("invalid number");
            }
            if (!Double.isFinite(value)) throw error("number is not finite");
            return t -> value;
        }

        private Node parseIdentifier() {
            int start = position;
            while (!atEnd()) {
                char value = source.charAt(position);
                if (!Character.isLetterOrDigit(value) && value != '_') break;
                position++;
            }
            String name = source.substring(start, position).toLowerCase(Locale.US);
            skipWhitespace();
            if (!take('(')) {
                switch (name) {
                    case "t": return t -> t;
                    case "pi": return t -> Math.PI;
                    case "e": return t -> Math.E;
                    default: throw error("unknown identifier " + name);
                }
            }

            List<Node> arguments = new ArrayList<>();
            skipWhitespace();
            if (!take(')')) {
                do {
                    arguments.add(parseAdditive());
                } while (take(','));
                require(')');
            }
            validateArity(name, arguments.size());
            return t -> evaluateFunction(name, arguments, t);
        }

        private void enter() {
            if (++depth > MAX_PARSE_DEPTH) throw error("expression is too deeply nested");
        }

        private void leave() {
            depth--;
        }

        private boolean take(char expected) {
            skipWhitespace();
            if (atEnd() || source.charAt(position) != expected) return false;
            position++;
            return true;
        }

        private void require(char expected) {
            if (!take(expected)) throw error("expected '" + expected + "'");
        }

        private void skipWhitespace() {
            while (!atEnd() && Character.isWhitespace(source.charAt(position))) position++;
        }

        private boolean atEnd() {
            return position >= source.length();
        }

        private IllegalArgumentException error(String message) {
            return new IllegalArgumentException(message + " at position " + position);
        }
    }

    private static void validateArity(String name, int count) {
        switch (name) {
            case "min":
            case "max":
                if (count >= 2) return;
                break;
            case "pow":
            case "atan2":
                if (count == 2) return;
                break;
            case "clamp":
                if (count == 3) return;
                break;
            case "sin":
            case "cos":
            case "tan":
            case "asin":
            case "acos":
            case "atan":
            case "sinh":
            case "cosh":
            case "tanh":
            case "sqrt":
            case "abs":
            case "floor":
            case "ceil":
            case "round":
            case "exp":
            case "ln":
            case "log":
            case "log10":
            case "sign":
            case "signum":
                if (count == 1) return;
                break;
            default:
                throw new IllegalArgumentException("unknown function " + name);
        }
        throw new IllegalArgumentException("invalid argument count for " + name);
    }

    private static double evaluateFunction(String name, List<Node> arguments, double t) {
        double first = arguments.isEmpty() ? 0.0 : arguments.get(0).evaluate(t);
        switch (name) {
            case "sin": return Math.sin(first);
            case "cos": return Math.cos(first);
            case "tan": return Math.tan(first);
            case "asin": return Math.asin(first);
            case "acos": return Math.acos(first);
            case "atan": return Math.atan(first);
            case "atan2": return Math.atan2(first, arguments.get(1).evaluate(t));
            case "sinh": return Math.sinh(first);
            case "cosh": return Math.cosh(first);
            case "tanh": return Math.tanh(first);
            case "sqrt": return Math.sqrt(first);
            case "abs": return Math.abs(first);
            case "floor": return Math.floor(first);
            case "ceil": return Math.ceil(first);
            case "round": return Math.rint(first);
            case "exp": return Math.exp(first);
            case "ln":
            case "log": return Math.log(first);
            case "log10": return Math.log10(first);
            case "sign":
            case "signum": return Math.signum(first);
            case "pow": return Math.pow(first, arguments.get(1).evaluate(t));
            case "clamp": {
                double low = arguments.get(1).evaluate(t);
                double high = arguments.get(2).evaluate(t);
                return Math.max(low, Math.min(high, first));
            }
            case "min": {
                double value = first;
                for (int index = 1; index < arguments.size(); index++) {
                    value = Math.min(value, arguments.get(index).evaluate(t));
                }
                return value;
            }
            case "max": {
                double value = first;
                for (int index = 1; index < arguments.size(); index++) {
                    value = Math.max(value, arguments.get(index).evaluate(t));
                }
                return value;
            }
            default: throw new IllegalArgumentException("unknown function " + name);
        }
    }
}

File: ios/PhiStudioIOS/PhiStudioIOS/Core/MathExpression.swift
import Foundation

public struct MathExpression: Sendable {
    private let root: Node

    public static func compile(_ source: String) throws -> MathExpression {
        var parser = Parser(source)
        let root = try parser.parse()
        return MathExpression(root: root)
    }

    public func evaluate(_ t: Double) throws -> Double {
        let value = root.evaluate(t)
        guard value.isFinite else {
            throw ChartError.invalidValue("Expression produced a non-finite value")
        }
        return value
    }

    private indirect enum Node: Sendable {
        case constant(Double)
        case variable
        case unary(Character, Node)
        case binary(Character, Node, Node)
        case function(String, [Node])

        func evaluate(_ t: Double) -> Double {
            switch self {
            case let .constant(value):
                return value
            case .variable:
                return t
            case let .unary(operation, operand):
                let value = operand.evaluate(t)
                return operation == "-" ? -value : value
            case let .binary(operation, left, right):
                let first = left.evaluate(t)
                let second = right.evaluate(t)
                switch operation {
                case "+": return first + second
                case "-": return first - second
                case "*": return first * second
                case "/": return first / second
                case "%": return first.truncatingRemainder(dividingBy: second)
                case "^": return pow(first, second)
                default: return .nan
                }
            case let .function(name, arguments):
                let values = arguments.map { $0.evaluate(t) }
                return Self.call(name, values)
            }
        }

        private static func call(_ name: String, _ values: [Double]) -> Double {
            let first = values.first ?? 0
            switch name {
            case "sin": return sin(first)
            case "cos": return cos(first)
            case "tan": return tan(first)
            case "asin": return asin(first)
            case "acos": return acos(first)
            case "atan": return atan(first)
            case "atan2": return atan2(first, values[1])
            case "sinh": return sinh(first)
            case "cosh": return cosh(first)
            case "tanh": return tanh(first)
            case "sqrt": return sqrt(first)
            case "abs": return abs(first)
            case "floor": return floor(first)
            case "ceil": return ceil(first)
            case "round": return first.rounded()
            case "exp": return exp(first)
            case "ln", "log": return log(first)
            case "log10": return log10(first)
            case "sign", "signum": return first == 0 ? 0 : (first > 0 ? 1 : -1)
            case "pow": return pow(first, values[1])
            case "clamp": return max(values[1], min(values[2], first))
            case "min": return values.min() ?? 0
            case "max": return values.max() ?? 0
            default: return .nan
            }
        }
    }

    private struct Parser {
        private static let maximumDepth = 64
        private var characters: [Character]
        private var position = 0
        private var depth = 0

        init(_ source: String) {
            characters = Array(source)
        }

        mutating func parse() throws -> Node {
            guard !characters.isEmpty else {
                throw error("Expression is required")
            }
            let node = try expression()
            skipWhitespace()
            guard atEnd else { throw error("Unexpected character") }
            return node
        }

        private mutating func expression() throws -> Node {
            try enter()
            defer { leave() }
            var node = try term()
            while true {
                if take("+") {
                    node = .binary("+", node, try term())
                } else if take("-") {
                    node = .binary("-", node, try term())
                } else {
                    return node
                }
            }
        }

        private mutating func term() throws -> Node {
            var node = try power()
            while true {
                if take("*") {
                    node = .binary("*", node, try power())
                } else if take("/") {
                    node = .binary("/", node, try power())
                } else if take("%") {
                    node = .binary("%", node, try power())
                } else {
                    return node
                }
            }
        }

        private mutating func power() throws -> Node {
            let node = try unary()
            return take("^") ? .binary("^", node, try power()) : node
        }

        private mutating func unary() throws -> Node {
            if take("+") { return .unary("+", try unary()) }
            if take("-") { return .unary("-", try unary()) }
            return try primary()
        }

        private mutating func primary() throws -> Node {
            skipWhitespace()
            if take("(") {
                let node = try expression()
                try require(")")
                return node
            }
            guard !atEnd else { throw error("Expected a value") }
            if current.isNumber || current == "." {
                return .constant(try number())
            }
            if current.isLetter || current == "_" {
                let name = identifier().lowercased()
                switch name {
                case "t", "x":
                    return .variable
                case "pi":
                    return .constant(.pi)
                case "e":
                    return .constant(exp(1))
                default:
                    try require("(")
                    var arguments: [Node] = []
                    if !take(")") {
                        repeat {
                            arguments.append(try expression())
                        } while take(",")
                        try require(")")
                    }
                    try Self.validate(name, count: arguments.count, error: error)
                    return .function(name, arguments)
                }
            }
            throw error("Expected a number, t, constant, function, or parenthesized expression")
        }

        private mutating func number() throws -> Double {
            skipWhitespace()
            let start = position
            var sawDigit = false
            while !atEnd, current.isNumber {
                sawDigit = true
                position += 1
            }
            if !atEnd, current == "." {
                position += 1
                while !atEnd, current.isNumber {
                    sawDigit = true
                    position += 1
                }
            }
            guard sawDigit else { throw error("Invalid number") }
            if !atEnd, (current == "e" || current == "E") {
                position += 1
                if !atEnd, (current == "+" || current == "-") { position += 1 }
                let exponentStart = position
                while !atEnd, current.isNumber { position += 1 }
                if exponentStart == position { throw error("Invalid exponent") }
            }
            let text = String(characters[start ..< position])
            guard let value = Double(text), value.isFinite else {
                throw error("Invalid number")
            }
            return value
        }

        private mutating func identifier() -> String {
            skipWhitespace()
            let start = position
            while !atEnd, (current.isLetter || current.isNumber || current == "_") {
                position += 1
            }
            return String(characters[start ..< position])
        }

        private static func validate(
            _ name: String,
            count: Int,
            error: (String) -> ChartError
        ) throws {
            switch name {
            case "min", "max":
                if count >= 2 { return }
            case "pow", "atan2":
                if count == 2 { return }
            case "clamp":
                if count == 3 { return }
            case "sin", "cos", "tan", "asin", "acos", "atan", "sinh", "cosh", "tanh",
                 "sqrt", "abs", "floor", "ceil", "round", "exp", "ln", "log", "log10",
                 "sign", "signum":
                if count == 1 { return }
            default:
                throw error("Unknown function \(name)")
            }
            throw error("Invalid argument count for \(name)")
        }

        private mutating func enter() throws {
            depth += 1
            if depth > Self.maximumDepth { throw error("Expression is too deeply nested") }
        }

        private mutating func leave() {
            depth -= 1
        }

        private mutating func take(_ expected: Character) -> Bool {
            skipWhitespace()
            guard !atEnd, current == expected else { return false }
            position += 1
            return true
        }

        private mutating func require(_ expected: Character) throws {
            guard take(expected) else { throw error("Expected '\(expected)'") }
        }

        private mutating func skipWhitespace() {
            while !atEnd, current.isWhitespace { position += 1 }
        }

        private var current: Character { characters[position] }
        private var atEnd: Bool { position >= characters.count }

        private func error(_ message: String) -> ChartError {
            .invalidValue("\(message) at position \(position)")
        }
    }
}
